Documentation (#10012)
GoogleCloudPlatform/magic-modules: mmv1/third_party/terraform/website/docs/r/composer_environment.html.markdown

  ```

* `image_version` -

>   (Required in Cloud Composer 2 and 3, optional in Cloud Composer 1)

Reviewer: ```suggestion
(Optional) If omitted, the default is the latest version of Composer 2.
```

>   **In Cloud Composer 2, you must specify an image with Cloud Composer 2**. Otherwise, the default image for Cloud Composer 1 is used. For more information about Cloud Composer images, see

Reviewer: ```suggestion
```

>   [Cloud Composer version list](https://cloud.google.com/composer/docs/concepts/versioning/composer-versions).

Reviewer: ```suggestion
```

The `workloads_config` block supports:

* `scheduler` -
  (Optional)

>   Configuration for resources used by Airflow schedulers.

Author: Made singular.

>   Configuration for resources used by Airflow schedulers.

Author: I checked with him, we use 'Airflow scheduler' in public docs.

* `triggerer` -
  (Optional)
  Configuration for resources used by Airflow triggerer.

* `web_server` -
  (Optional)
  Configuration for resources used by Airflow web server.

* `worker` -
  (Optional)
  Configuration for resources used by Airflow workers.

> 

Author: This is Composer 2 section. There's dag_processor in Composer 3 section.

  Configuration options for storage used by Composer environment. Structure is documented below.

The `config` block supports:

* `node_config` -
  (Optional)
  The configuration used for the Kubernetes Engine cluster. Structure is documented below.

* `software_config` -
  (Optional)
  The configuration settings for software (Airflow) inside the environment. Structure is
  documented below.

* `enable_private_environment` -
  (Optional, [Beta](https://terraform.io/docs/providers/google/guides/provider_versions.html), Cloud Composer 3 only)
  If true, a private Composer environment will be created.

* `enable_private_builds_only` -
  (Optional, [Beta](https://terraform.io/docs/providers/google/guides/provider_versions.html), Cloud Composer 3 only)
  If true, builds performed during operations that install Python packages have only private connectivity to Google services.
  If false, the builds also have access to the internet.

* `encryption_config` -
  (Optional)
  The encryption options for the Cloud Composer environment and its
  dependencies.

* `maintenance_window` -
  (Optional)
  The configuration settings for Cloud Composer maintenance windows.

* `workloads_config` -

>   (Optional, Cloud Composer 2 only)

Author: Removed from both, Composer 2 and 3 section.

  The Kubernetes workloads configuration for GKE cluster associated with the
  Cloud Composer environment.

* `environment_size` -
  (Optional, Cloud Composer 2 only)
  The environment size controls the performance parameters of the managed
  Cloud Composer infrastructure that includes the Airflow database. Values for
  environment size are `ENVIRONMENT_SIZE_SMALL`, `ENVIRONMENT_SIZE_MEDIUM`,
  and `ENVIRONMENT_SIZE_LARGE`.

> * `resilience_mode` -

Author: Removed for now.

* `composer_network_attachment` -
  (Optional, [Beta](https://terraform.io/docs/providers/google/guides/provider_versions.html), Cloud Composer 3 only)
  PSC (Private Service Connect) Network entry point. Customers can pre-create the Network Attachment 
  and point Cloud Composer environment to use. It is possible to share network attachment among many environments, 
  provided enough IP addresses are available.

* `service_account` -
  (Optional)
  The Google Cloud Platform Service Account to be used by the
  node VMs. If a service account is not specified, the "default"
  Compute Engine service account is used. Cannot be updated. If given,
  note that the service account must have `roles/composer.worker`
  for any GCP resources created under the Cloud Composer Environment.

* `tags` -
  (Optional)
  The list of instance tags applied to all node VMs. Tags are
  used to identify valid sources or targets for network
  firewalls. Each tag within the list must comply with RFC1035.
  Cannot be updated.

* `composer_internal_ipv4_cidr_block` -
  (Optional, [Beta](https://terraform.io/docs/providers/google/guides/provider_versions.html), Cloud Composer 3 only)

>   At least /20 IPv4 cidr range that will be used by Composer internal components.

Author: done

  and cannot be overridden.

* `pypi_packages` -
  (Optional)
  Custom Python Package Index (PyPI) packages to be installed
  in the environment. Keys refer to the lowercase package name (e.g. "numpy"). Values are the lowercase extras and
  version specifier (e.g. "==1.12.0", "[devel,gcp_api]", "[devel]>=1.8.2, <1.9.2"). To specify a package without
  pinning it to a version specifier, use the empty string as the value.

* `env_variables` -
  (Optional)
  Additional environment variables to provide to the Apache Airflow scheduler, worker, and webserver processes.
  Environment variable names must match the regular expression `[a-zA-Z_][a-zA-Z0-9_]*`.
  They cannot specify Apache Airflow software configuration overrides (they cannot match the regular expression
  `AIRFLOW__[A-Z0-9_]+__[A-Z0-9_]+`), and they cannot match any of the following reserved names:
  ```
  AIRFLOW_HOME
  C_FORCE_ROOT
  CONTAINER_NAME
  DAGS_FOLDER
  GCP_PROJECT
  GCS_BUCKET
  GKE_CLUSTER_NAME
  SQL_DATABASE
  SQL_INSTANCE
  SQL_PASSWORD
  SQL_PROJECT
  SQL_REGION
  SQL_USER
  ```

* `image_version` -

>   (Required in Cloud Composer 2, optional in Cloud Composer 1)

Author: Replaced with the description of Composer 3 versioning.

* `enabled` -
  (Required)
  Whether or not Cloud Data Lineage integration is enabled.

<a name="nested_encryption_config_comp_2"></a>The `encryption_config` block supports:

* `kms_key_name` -
  (Required)
  Customer-managed Encryption Key available through Google's Key Management Service. It must
  be the fully qualified resource name,
  i.e. projects/project-id/locations/location/keyRings/keyring/cryptoKeys/key. Cannot be updated.

<a name="nested_maintenance_window_comp_2"></a>The `maintenance_window` block supports:

* `start_time` -
  (Required)
  Start time of the first recurrence of the maintenance window.

* `end_time` -
  (Required)
  Maintenance window end time. It is used only to calculate the duration of the maintenance window.
  The value for end-time must be in the future, relative to 'start_time'.

* `recurrence` -
  (Required)
  Maintenance window recurrence. Format is a subset of RFC-5545 (https://tools.ietf.org/html/rfc5545) 'RRULE'.
  The only allowed values for 'FREQ' field are 'FREQ=DAILY' and 'FREQ=WEEKLY;BYDAY=...'.
  Example values: 'FREQ=WEEKLY;BYDAY=TU,WE', 'FREQ=DAILY'.

The `recovery_config` block supports:

> * `scheduled_snapshots_config` -

Author: Removed for now.

  (Optional)
  Configuration for resources used by Airflow workers.

The `scheduler` block supports:

* `cpu` -
  (Optional)
  The number of CPUs for a single Airflow scheduler.

* `memory_gb` -
  (Optional)
  The amount of memory (GB) for a single Airflow scheduler.

* `storage_gb` -
  (Optional)
  The amount of storage (GB) for a single Airflow scheduler.

* `count` -
  (Optional)
  The number of schedulers.

The `triggerer` block supports:

* `cpu` -
  (Required)
  The number of CPUs for a single Airflow triggerer.

* `memory_gb` -
  (Required)
  The amount of memory (GB) for a single Airflow triggerer.

* `count` -
  (Required)
  The number of Airflow triggerers.

The `web_server` block supports:

* `cpu` -
  (Optional)
  The number of CPUs for the Airflow web server.

* `memory_gb` -
  (Optional)
  The amount of memory (GB) for the Airflow web server.

* `storage_gb` -
  (Optional)
  The amount of storage (GB) for the Airflow web server.

The `worker` block supports:

* `cpu` -
  (Optional)
  The number of CPUs for a single Airflow worker.

* `memory_gb` -
  (Optional)
  The amount of memory (GB) for a single Airflow worker.

> * `float storage_gb`

Reviewer 2: ```suggestion
* `storage_gb`
```

  (Optional)
  The amount of storage (GB) for a single Airflow worker.

* `min_count` -
  (Optional)
  The minimum number of Airflow workers that the environment can run. The number of workers in the
  environment does not go above this number, even if a lower number of workers can handle the load.

* `max_count` -
  (Optional)
  The maximum number of Airflow workers that the environment can run. The number of workers in the
  environment does not go above this number, even if a higher number of workers is required to
  handle the load.

## Argument Reference - Cloud Composer 3

> **Please note: This documentation corresponds to Composer 3, which is not yet released.**

Reviewer 2: What does "not yet released" mean in this context? We have tests for composer 3 already running successfully in our nightly builds.

  and versions of dependencies. To distinguish between these versions in Composer 3, you can optionally specify a
  build number to pin to a specific Airflow release.
  Example: composer-3-airflow-2.6.3-build.4.

  The image version in Composer 3 must match the regular expression:
  `composer-(([0-9]+)(\.[0-9]+\.[0-9]+(-preview\.[0-9]+)?)?|latest)-airflow-(([0-9]+)((\.[0-9]+)(\.[0-9]+)?)?(-build\.[0-9]+)?)`
  Example: composer-3-airflow-2.6.3-build.4

  **Important**: In-place upgrade for Composer 3 is not yet supported.

* `cloud_data_lineage_integration` -
  (Optional, [Beta](https://terraform.io/docs/providers/google/guides/provider_versions.html),
  Cloud Composer environments in versions composer-2.1.2-airflow-*.*.* and later)
  The configuration for Cloud Data Lineage integration. Structure is
  [documented below](#nested_cloud_data_lineage_integration).

* `web_server_plugins_mode` -
  (Optional, [Beta](https://terraform.io/docs/providers/google/guides/provider_versions.html), Cloud Composer 3 only)
  Web server plugins configuration. Can be either 'ENABLED' or 'DISABLED'. Defaults to 'ENABLED'.

> <a name="nested_cloud_data_lineage_integration"></a>The `cloud_data_lineage_integration` block supports:

Reviewer 2: All the blocks like this will need to be disambiguated to specify whether they are composer 1, composer 2, or composer 3 - otherwise links will always take you to the composer 1 version.

Author: done.

  If true, a private Composer environment will be created.

* `enable_private_builds_only` -
  (Optional, [Beta](https://terraform.io/docs/providers/google/guides/provider_versions.html), Cloud Composer 3 only)
  If true, builds performed during operations that install Python packages have only private connectivity to Google services.
  If false, the builds also have access to the internet.

* `encryption_config` -
  (Optional)
  The encryption options for the Cloud Composer environment and its
  dependencies.

* `maintenance_window` -
  (Optional)
  The configuration settings for Cloud Composer maintenance windows.

* `workloads_config` -
  (Optional)
  The Kubernetes workloads configuration for GKE cluster associated with the
  Cloud Composer environment.

* `environment_size` -
  (Optional)
  The environment size controls the performance parameters of the managed
  Cloud Composer infrastructure that includes the Airflow database. Values for
  environment size are `ENVIRONMENT_SIZE_SMALL`, `ENVIRONMENT_SIZE_MEDIUM`,
  and `ENVIRONMENT_SIZE_LARGE`.

* `data_retention_config` -

>   (Optional, Cloud Composer 2.0.23 or newer only)

Reviewer: ```suggestion
  (Optional, Cloud Composer 2.0.23 or later only)
```

>   Configuration setting for airflow data rentention mechanism. Structure is

Reviewer: ```suggestion
  Configuration setting for Airflow database retention mechanism. Structure is
```

  [documented below](#nested_data_retention_config).

<a name="nested_data_retention_config"></a>The `data_retention_config` block supports:
* `task_logs_retention_config` - 
  (Optional)

>   The configuration setting for Task Logs. Structure is

Reviewer: ```suggestion
  The configuration setting for Airflow task logs. Structure is
```

  and cannot be overridden.

* `pypi_packages` -
  (Optional)
  Custom Python Package Index (PyPI) packages to be installed
  in the environment. Keys refer to the lowercase package name (e.g. "numpy"). Values are the lowercase extras and
  version specifier (e.g. "==1.12.0", "[devel,gcp_api]", "[devel]>=1.8.2, <1.9.2"). To specify a package without
  pinning it to a version specifier, use the empty string as the value.

* `env_variables` -
  (Optional)
  Additional environment variables to provide to the Apache Airflow scheduler, worker, and webserver processes.
  Environment variable names must match the regular expression `[a-zA-Z_][a-zA-Z0-9_]*`.
  They cannot specify Apache Airflow software configuration overrides (they cannot match the regular expression
  `AIRFLOW__[A-Z0-9_]+__[A-Z0-9_]+`), and they cannot match any of the following reserved names:
  ```
  AIRFLOW_HOME
  C_FORCE_ROOT
  CONTAINER_NAME
  DAGS_FOLDER
  GCP_PROJECT
  GCS_BUCKET
  GKE_CLUSTER_NAME
  SQL_DATABASE
  SQL_INSTANCE
  SQL_PASSWORD
  SQL_PROJECT
  SQL_REGION
  SQL_USER
  ```

* `image_version` -

>   (Required in Cloud Composer 2 and 3, optional in Cloud Composer 1)

Reviewer 2: +1

>   (Required in Cloud Composer 2 and 3, optional in Cloud Composer 1)

Author: Default version is still Composer 2.

  In Cloud Composer 3, you can only specify 3 in Cloud Composer portion of the image version.

>   Environment details will only show 3 as the Composer version as well.

Reviewer: ```suggestion
```

  The Apache Airflow portion of the image version is a full semantic version that points to one of the
  supported Apache Airflow versions, or an alias in the form of only major, major.minor or major.minor.patch versions specified.
  Like in Composer 1 and 2, a given Airflow version is released multiple times in Composer, with different patches
  and versions of dependencies. To distinguish between these versions in Composer 3, you can optionally specify a

>   build number to pin to a specific Airflow release. Environment details will always show the Airflow build number.

Reviewer: ```suggestion
  build number to pin to a specific Airflow release.
```

>   Example: 2.6.3-build.4.

Reviewer: ```suggestion
  Example: composer-3-airflow-2.6.3-build.4.
```

  The image version in Composer 3 must match the regular expression:
  `composer-(([0-9]+)(\.[0-9]+\.[0-9]+(-preview\.[0-9]+)?)?|latest)-airflow-(([0-9]+)((\.[0-9]+)(\.[0-9]+)?)?(-build\.[0-9]+)?)`
  Example: composer-3-airflow-2.6.3-build.4

  **Important**: In-place upgrade for Composer 3 is not yet supported.

* `cloud_data_lineage_integration` -
  (Optional, [Beta](https://terraform.io/docs/providers/google/guides/provider_versions.html),

>   Cloud Composer environments in versions composer-2.1.2-airflow-*.*.* and newer)

Reviewer: ```suggestion
  Cloud Composer environments in versions composer-2.1.2-airflow-*.*.* and later)
```

  The configuration for Cloud Data Lineage integration. Structure is
  [documented below](#nested_cloud_data_lineage_integration).

* `web_server_plugins_mode` -
  (Optional, [Beta](https://terraform.io/docs/providers/google/guides/provider_versions.html), Cloud Composer 3 only)

>   Web server plugins configuration. Should be either 'ENABLED' or 'DISABLED'. Defaults to 'ENABLED'.

Reviewer: ```suggestion
  Web server plugins configuration. Can be either 'ENABLED' or 'DISABLED'. Defaults to 'ENABLED'.
```
